File: tests/cs-sdk/TestSignalSubscribeIntegration.cs
/*
 * KnotLink SDK - C# Integration Test
 * Signal-Subscribe pattern: Subscriber + Sender
 */

using System;
using System.Threading;
using System.Threading.Tasks;
using KnotLink;

internal static class TestSignalSubscribeIntegration
{
    private static readonly CancellationTokenSource Cts = new();

    public static async Task Main()
    {
        var t = Task.Run(RunSubscriber);
        await Task.Delay(100);
        await RunSender();
        Cts.Cancel();
        await t;
    }

    private static async Task RunSubscriber()
    {
        var sub = new SignalSubscriber("app.knotlinksdk.test", "test_signal")
        {
            OnSignalAsync = data =>
            {
                Console.WriteLine($"[Subscriber] Received: {data}");
                return Task.CompletedTask;
            }
        };

        Console.WriteLine("[Subscriber] Listening...");
        while (!Cts.IsCancellationRequested)
        {
            await Task.Delay(1000, Cts.Token);
        }
    }

    private static async Task RunSender()
    {
        await Task.Delay(1000);
        var sender = new SignalSender("app.knotlinksdk.test", "test_signal");
        try
        {
            await sender.EmitAsync("Hello from Sender!");
            Console.WriteLine("[Sender] Signal sent.");
        }
        finally
        {
            sender.Dispose();
        }
    }
}

File: knotlink-sdk-qt/tests/test_signal_subscribe_integration.cpp
/*
 * KnotLink SDK - Qt C++ Integration Test
 * Signal-Subscribe pattern: Subscriber + Sender
 */

#include <QCoreApplication>
#include <QThread>
#include <QTimer>
#include <QDebug>
#include "../knotlink/signalsubscriber.h"
#include "../knotlink/signalsender.h"

static bool shutdown = false;

void run_subscriber() {
    SignalSubscriber sub("app.knotlinksdk.test", "test_signal");
    QObject::connect(&sub, &SignalSubscriber::receivedData,
        [](const QString &data) {
            qDebug() << "[Subscriber] Received:" << data;
        });
    qDebug() << "[Subscriber] Listening...";
    while (!shutdown) {
        QThread::msleep(1000);
    }
}

void run_sender() {
    QThread::msleep(1000);
    SignalSender sender;
    sender.setConfig("app.knotlinksdk.test", "test_signal");
    sender.emitt("Hello from Sender!");
    qDebug() << "[Sender] Signal sent.";
    QThread::msleep(2000);
    shutdown = true;
}

int main(int argc, char *argv[]) {
    QCoreApplication app(argc, argv);

    QThread* t = QThread::create(run_subscriber);
    t->start();

    QTimer::singleShot(100, run_sender);
    QTimer::singleShot(5000, &app, &QCoreApplication::quit);

    app.exec();
    t->wait();
    return 0;
}
